ct: Refactor data plane API

Lazin opened the pull request:
Fix few minor problem in L0 (namespaces, include paths don't match locations).
Also, make `impl` class in `data_plane_impl` use `ssx::sharded_service_container` to shard all pipeline components. This is needed to allow the followup PR to introduce some cross shard functionality in L0.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

Lazin committed:
ct: Add two-step shutdown to pipe

Pipeline components are supposed to use abort source from the underlying
pipeline. The idea is that the pipeline is a source of the data for
components and it also controls when they should stop. This allows us to
avoid passing abort source into many different places which share the
pipeline.

The problem is that it doesn't work well with
'ssx::sharded_service_container'. If the pipeline and all components are
services the container expects them to stop in certain order and
naturally, the pipeline is stopped last (because it's created first).
This makes 'stop' method of the pipeline component to get stuck.

This commit adds a 'shutdown' method. The methods triggers shutdown for
all components which use pipeline's abort source. The second step is a
'stop' method that works as before. The 'shutdown' call is optional.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Add cluster_services_impl to L0

The class implements cluster_services interfaces and is supposed to be
used with the batcher.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Add c-tor to batch_cache

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Refactor data_plane_impl

Make 'impl' class use the 'ssx::sharded_service_container'. Make
pipelines and their components to be sharded services.

The motivation here is that we want some cross-shard functionality in L0
(aggregated upload per broker and not per shard) which will require us
to create a peering sharded service. But with the current approach this
is impossible. The whole 'impl' is a monolitic object which is 'sharded'
through a wrapper (state_accessors).

This commit also changes 'state_accessor' to be non-owning. It stores a
ref to the 'data_plane_api' object. In the current form its only purpose
is to pass state around without introducing dependencies.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Split ct/level_zero/reader

Extract top level L0 reader into its own module.
Currently, the 'reader' module contains both pipeline component that
fetches data for the read path and the actual L0 read path (log reader).
Thie reader is used from the kafka/data layer through the ct/frontend
module. This commit fixes this by extracting top level L0 reader into
the fe_reader module.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--02853f2e-885c-4964-9825-903a6558836c-->
<!--TEST RESULTS-->
<!--4b15f1b4-31e5-4e85-8eea-1642deb32304-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70640">build#70640</a></summary>


| test_class                 | test_method                  | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------|:-----------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeClusterRestoreTest | test_slow_tiered_storage_dlq | {"catalog_type": "nessie", "cloud_storage_type": 1}                                                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/70640#01989e47-3a96-48f0-9cb3-a53ce066d0ad | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest   | test_node_operations         | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70640#01989e4b-780a-4426-bc5c-0e484d2a6ff2 | FLAKY         | 20/21    | upstream reliability is '99.48717948717949'. current run reliability is '95.23809523809523'. drift is 4.24908 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70708">build#70708</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70708#0198a2de-91f3-422e-8561-fa7d387cc219 | FLAKY         | 17/21    | upstream reliability is '93.59477124183006'. current run reliability is '80.95238095238095'. drift is 12.64239 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat commented on src/v/cloud_topics/level_zero/throttler/throttler.h:
```diff
@@ -38,6 +38,8 @@ class throttler {
 public:
     // TODO: add config properties for limits
     explicit throttler(size_t tput_limit, l0::write_pipeline<Clock>::stage s);
+    explicit throttler(
+      size_t tput_limit, ss::sharded<l0::write_pipeline<Clock>>& pipeline);
```
definition is missing in this commit

Lazin replied:
moved this to the proper commit

dotnwat commented on src/v/cloud_topics/frontend/BUILD:
```diff
@@ -35,7 +35,7 @@ redpanda_cc_library(
         "//src/v/cloud_topics:data_plane_api",
         "//src/v/cloud_topics/level_zero/common:extent_meta",
-        "//src/v/cloud_topics/level_zero/reader",
+        "//src/v/cloud_topics/level_zero/fe_reader",
```
maybe let's change `fe` to be `frontend`?

Lazin replied:
that'd be third thing called 'frontend' in the CT directory structure, I'd prefer to keep it like that

rockwotj replied:
No he is saying expand the abbreviation.

Make it `frontend_reader`

Lazin replied:
I assumed that this is a nit because of "maybe"

Lazin replied:
renamed it to 'frontend_reader'

dotnwat reviewed (commented):
lgtm. just a couple minor things

rockwotj commented on src/v/cloud_topics/app.h:
```diff
@@ -66,6 +67,7 @@ class app : public ssx::sharded_service_container {
     // TODO: add 'get_control_plane_api' etc
 
 private:
+    ss::shared_ptr<data_plane_api> data_plane;
```
why is this `shared_ptr` and not `std::unique_ptr`?

Lazin replied:
fixed that

rockwotj reviewed (commented):
LGTM, but one question to add to Noah's.

Lazin committed:
ct: Fix include paths and namespaces

Fix includes in the ct/level_zero/pipeline. The include_prefix was set
to 'cloud_topics/level_zero' instead of
'cloud_topics/level_zero/pipline'. Because of that the headers from the
'pipeline' submodule were included as 'cloud_topics/level_zero/*.h'
instead of 'cloud_topics/level_zero/pipeline/*.h'.

This commit changes the 'include_path' and updates include directives.
It also fixes another issue. The 'l0' namespace is used inconsistently.
This commit fixes that by adding l0 namespace to batcher, throttler,
etc.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

rockwotj reviewed (dismissed):
Thanks!

dotnwat commented:
> ERROR: no such package 'src/v/cloud_topics/level_zero/fe_reader': BUILD file not found in any > of the following directories. Add a BUILD file to a directory to mark it as a package.
> --
>  | 2025-08-12 17:16:07 PDT | - src/v/cloud_topics/level_zero/fe_reader

dotnwat reviewed (dismissed):
lgtm just got that build error

andrwng commented on src/v/cloud_topics/data_plane_impl.cc:
```diff
@@ -31,59 +34,75 @@ namespace experimental::cloud_topics {
 public:
-    impl(
+    impl()
+      : ssx::sharded_service_container("cloud_topics::data_plane_impl") {}
```
Unfortunately, this should take in a string to uniquely identify the logger. Otherwise tests that use multiple nodes in the same cluster will crash on this (see https://github.com/redpanda-data/redpanda/pull/27200 for a similar change)

Lazin replied:
fixed, thank
